<!DOCTYPE html>
<html>

<head>
    <title>Make spatial matrix deconvolution</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="https://gcmodeller.org/lib/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>deconv_spatial {Erica}</td>
                <td style="text-align: right;">R Documentation</td>
            </tr>
        </table>

        <h2>Make spatial matrix deconvolution</h2>

        <h3>Description</h3>

        <p>Reference-free cell-type deconvolution of pixel-resolution spatially resolved transcriptomics data</p>

        <h3>Usage</h3>

        <div style="width: 100%; padding-left: 3%; color: black;">
            <pre><code class='language-R' style="color: black; font-weight:normal;"><strong>deconv_spatial</strong>(<i>expr.mat</i>,
    <i>n.layers</i> = <span style='color: green;'>4</span>,
    <i>top.genes</i> = <span style='color: green;'>1000</span>,
    <i>alpha</i> = <span style='color: green;'>2</span>,
    <i>beta</i> = <span style='color: green;'>0.5</span>,
    <i>iteration</i> = <span style='color: green;'>150</span>,
    <i>prefix</i> = <span style='color: brown;'><strong>'class'</strong></span>,
    <i>make.gene.filters</i> = <span style='color: blue !important;'>TRUE</span>,
    <i>filter.range</i> = [<span style='color: green;'>0.95</span>],
    <i>unify.scale</i> = <span style='color: green;'>10</span>,
    <i>log.norm</i> = <span style='color: blue !important;'>TRUE</span>,
    <i>n.threads</i> = <span style='color: blue !important;'>NULL</span>);</code></pre>
        </div>

        <h3>Arguments</h3>

        <table>
            
<dt>expr.mat</dt>
<dd><p><p>a dataframe object or a GCModeller expression matrix object, the matrix
object should be in format of spatial spot in rows and the gene features in
columns.</p></p></dd>


<dt>n.layers</dt>
<dd><p><p>an integer value for specific the cell layer class number</p></p></dd>


<dt>top.genes</dt>
<dd><p><p>set the number of top genes when export the expression matrix
in each cell layers</p></p></dd>


<dt>n.threads</dt>
<dd><p><p>default null means controls the thread number for parallel sampling via
the r environment config: <code>getOption("n_threads")</code>.</p></p></dd>

        </table>


        <h3 style="display: block;">Details</h3>

        <p>the generated expression matrix object could be save as csv matrix by use
the function <code>geneExpression::write.expr_matrix</code> from the GCModeller package.</p>

        <h3 style="display: block">Authors</h3>

        xieguigang

        <h3>Value</h3>

        <p>A tuple list object that contains multiple result object inside:
<ol>
<li>single_cells: an expression matrix object that contains the single cell data</li>
<li>deconv_spatial: an expression matrix object that contains the cell</li>
</ol></p>

<p>layer deconvolution result data
<ol>
<li>cell_layers: the cell layer composition data</li>
<li>gibbs_LDA: the result of the LDA gibbs sampling result outputs, the top</li>
</ol></p>

<p>features in each cell class</p>

        <h3 style="display: none;">Examples</h3>

        

        <hr />
        <div style="text-align: center;">[Package <em>Erica</em> version 0.1.0 <a
                href="../Erica.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="https://gcmodeller.org/lib/R_syntax.js"></script>
<script type="text/javascript" src="https://gcmodeller.org/lib/highlights.js"></script>
<script type="text/javascript">r_highlights("example_r");</script>

</html>
